Fix :fullscreen documentation

yookoala opened the pull request:
### Description
Fix both the example and description of the :fullscreen pseudo class page.

### Motivation

Problems:
* The example doesn't work without Javascript but the script is not provided.
* The description of :fullscreen pseudo class is wrong. The pseudo class does not activate when the browser or document is in fullscreen mode at all. It matches **element** that is fullscreen.

Fixes:
* Add javascript and a parent element for the Fullscreen API to work with. Parent element is not required but it better demonstrates the nature of the whole API set.
* Correct the descriptions.
* Use color name instead of rgb to improve readability of the example.


### Additional details

From the specification of :fullscreen, [5.1](https://fullscreen.spec.whatwg.org/#:fullscreen-pseudo-class):

<blockquote>
The :fullscreen pseudo-class must match any [element](https://dom.spec.whatwg.org/#concept-element) for which one of the following conditions is true:

*  element’s [fullscreen flag](https://fullscreen.spec.whatwg.org/#fullscreen-flag) is set.

* element is a [shadow host](https://dom.spec.whatwg.org/#element-shadow-host) and the result of [retargeting](https://dom.spec.whatwg.org/#retarget) its [node document](https://dom.spec.whatwg.org/#concept-node-document)’s [fullscreen element](https://fullscreen.spec.whatwg.org/#fullscreen-element) against element is element.
</blockquote>

A JSFiddle example showing the working example: https://jsfiddle.net/yookoala/veu1n3rt/

### Related issues and pull requests

Fixes #29950

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

yookoala committed:
Fix :fullscreen documentation

Problems:
* The example doesn't work without Javascript but the script is not provided.
* The description of :fullscreen pseudo class is wrong. The pseudo class does not activate when the browser or document is in fullscreen mode at all. It matches **element** that is fullscreen.

Fixes:
* Add javascript and a parent element for the Fullscreen API to work with. Parent element is not required but it better demonstrates the nature of the whole API set.
* Correct the descriptions.
* Use color name instead of rgb to improve readability of the example.

yookoala committed:
Update files/en-us/web/css/_colon_fullscreen/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Update files/en-us/web/css/_colon_fullscreen/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Update files/en-us/web/css/_colon_fullscreen/index.md

Fix typo

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala commented:
Note: I've consider to create a live example for this, but I can't seem to trigger fullscreen API from within the iframe of live example. Please advice if there are better ways to demonstrate this with live demo. Thanks.

chrisdavidmills commented:
> Note: I've consider to create a live example for this, but I can't seem to trigger fullscreen API from within the iframe of live example. Please advice if there are better ways to demonstrate this with live demo. Thanks.

@yookoala hi there!

Before I do the editorial review, I've decided to address the live example issue. I wasn't sure if there was some issue with triggering fullscreen mode from an `<iframe>`, or something. Other fullscreen demos on MDN link to external examples rather than using in-page live examples as well...

But good news — I got the following live example markdown working on my local content repo. Try it out on yours and see what you think: https://pastebin.com/XWLKCwmr

(I have put it in a pastebin so you can easily see the raw markdown, without GitHub interpreting it in the comment).

yookoala committed:
Simplified example CSS

Simplified the example CSS rules and explainations to better convey the feature of the selector.

yookoala committed:
Update files/en-us/web/css/_colon_fullscreen/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Update files/en-us/web/css/_colon_fullscreen/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

chrisdavidmills commented on files/en-us/web/css/_colon_fullscreen/index.md:
```diff
@@ -23,41 +23,62 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
 ## Examples
 
-In this example, the color of a button is changed depending on whether or not the document is in fullscreen mode. This is done without needing to specifically apply style changes using JavaScript.
+In this example, the color of a button is changed depending on whether or not it's parent div is in fullscreen mode. This is done without needing to specifically apply style changes using JavaScript.
```
I think the example is a bit weird — why don't we just change the parent `<div>`'s background color in fullscreen mode, so it is more obvious what is going on? Changing the color of the control button seems like a strange choice.

Also, I don't think the last sentence here is needed; the whole point of a pseudo-class is that it allows style updates on a state change, without needing JS.

In which case this intro paragraph would turn into:

> In this example, the background color of a `<div>` is changed depending on whether or not it is in fullscreen mode.

yookoala replied:
That is actually a good suggestion. Will do.

yookoala commented:
> > Note: I've consider to create a live example for this, but I can't seem to trigger fullscreen API from within the iframe of live example. Please advice if there are better ways to demonstrate this with live demo. Thanks.
> 
> @yookoala hi there!
> 
> Before I do the editorial review, I've decided to address the live example issue. I wasn't sure if there was some issue with triggering fullscreen mode from an `<iframe>`, or something. Other fullscreen demos on MDN link to external examples rather than using in-page live examples as well...
> 
> But good news — I got the following live example markdown working on my local content repo. Try it out on yours and see what you think: https://pastebin.com/XWLKCwmr
> 
> (I have put it in a pastebin so you can easily see the raw markdown, without GitHub interpreting it in the comment).

Hello. Thanks for the reply.

Sorry to need to ask a basic question. How do I test your pastebin code with this repo? Do I need to locally clone and build the whole site?

chrisdavidmills commented on files/en-us/web/css/_colon_fullscreen/index.md:
````diff
@@ -23,41 +23,62 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
 
 ```html
-<h1>MDN Web Docs Demo: :fullscreen pseudo-class</h1>
+<div style="background-color: lightyellow">
````
If you do agree with my suggestion about the example, then we could get rid of this inline style, which generally isn't considered to be a good practice.

```suggestion
<div>
```

yookoala replied:
Sure.

chrisdavidmills commented on files/en-us/web/css/_colon_fullscreen/index.md:
```diff
@@ -23,41 +23,62 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
-  entirely using CSS.
-</p>
+  <p>
+    This demo uses the <code>:fullscreen</code> pseudo-class to automatically
```
This description would also need updating...

yookoala replied:
Sure.

chrisdavidmills commented on files/en-us/web/css/_colon_fullscreen/index.md:
````diff
@@ -23,41 +23,62 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
 ```
 
-The {{HTMLElement("button")}} with the ID `"fs-toggle"` will change between pale red and pale green depending on whether or not the document is in fullscreen mode.
+The {{HTMLElement("button")}} with the ID `"fs-toggle"` will change between light pink and light green depending on whether or not the parent div is in fullscreen mode.
````
I'd suggest that you put this paragraph just above the live sample macro call, then you implement it as per my other comment, and change it to:

```suggestion
The `<div>` will change between light yellow and light pink depending on whether or not the parent div is in fullscreen mode — try it out.
```

chrisdavidmills commented on files/en-us/web/css/_colon_fullscreen/index.md:
````diff
@@ -23,41 +23,62 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
 ```css
-#fs-toggle:not(:fullscreen) {
-  background-color: #afa;
+div:not(:fullscreen) #fs-toggle {
````
This CSS is overcomplicated and confusing. The use of `:not` is clever, but it makes the effect of `:fullscreen` less obvious. Rather than selecting for an element when it is explicitly not in fullscreen, you could just select the element. For example, the code in my update suggestion would look like this:

```css
div {
  background-color: lightyellow;
}

div:fullscreen {
    background-color: lightpink;
}
```

chrisdavidmills commented on files/en-us/web/css/_colon_fullscreen/index.md:
````diff
@@ -23,41 +23,62 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
 ```
 
-When the document _is_ in fullscreen mode, the following CSS applies instead, setting the background color to a pale shade of red.
+When the parent div _is_ in fullscreen mode, the following CSS applies instead, setting the background color to light pink.
````
```suggestion
When the parent div is in fullscreen mode, the following CSS applies instead, setting its background color to light pink.
```

chrisdavidmills reviewed (changes requested):
@yookoala nice work, this is definitely an improvement. I think it will be even better if you install a live example as shown in my other comment, and change the example as suggested.

Thanks!

chrisdavidmills commented:
> Sorry to need to ask a basic question. How do I test your pastebin code with this repo? Do I need to locally clone and build the whole site?

@yookoala no problem.

So the best way to do it would be to clone and build the site locally, yes:

1. Fork https://github.com/mdn/content and clone it locally.
2. build it as indicated at https://github.com/mdn/content#build-the-site
3. take any markdown source page and paste my code in, then start the local server with `yarn start` and navigate to the page you have edited.

It is pretty easy and painless to get MDN running locally, and very useful to get set up if you intend to make more contributions to MDN in the future. And it is fun having it running in localhost!

However, if you don't want to spend the time doing that, I have another suggestion for you — you could just make your changes in the PR and submit them, and then the dynamic preview build (see https://github.com/mdn/content/pull/29951#issuecomment-1788528237) will update to show how your edit renders.

yookoala commented:
> > Sorry to need to ask a basic question. How do I test your pastebin code with this repo? Do I need to locally clone and build the whole site?
> 
> @yookoala no problem.
> 
> So the best way to do it would be to clone and build the site locally, yes:
> 
>     1. Fork https://github.com/mdn/content and clone it locally.
> 
>     2. build it as indicated at https://github.com/mdn/content#build-the-site
> 
>     3. take any markdown source page and paste my code in, then start the local server with `yarn start` and navigate to the page you have edited.
> 
> 
> It is pretty easy and painless to get MDN running locally, and very useful to get set up if you intend to make more contributions to MDN in the future. And it is fun having it running in localhost!
> 
> However, if you don't want to spend the time doing that, I have another suggestion for you — you could just make your changes in the PR and submit them, and then the dynamic preview build (see [#29951 (comment)](https://github.com/mdn/content/pull/29951#issuecomment-1788528237)) will update to show how your edit renders.

Is there any way to partially build the site? It takes a very long time to build the whole site once, but what I need is only 1 single page.

chrisdavidmills commented:
> Is there any way to partially build the site? It takes a very long time to build the whole site once, but what I need is only 1 single page.

@yookoala I don't think there is, no.m The good news is that once you've done the initial build, it is extremely fast at rendering updates.

yookoala committed:
Further simplified as live example

* Reduce description text.
* Simplifed code.
* Make the example a live example.

yookoala committed:
Follow indentation rules

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Follow indentation rules

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Follow indentation rules

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Follow indentation rules

Update files/en-us/web/css/_colon_fullscreen/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Follow indentation rules

Update files/en-us/web/css/_colon_fullscreen/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala committed:
Update files/en-us/web/css/_colon_fullscreen/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

yookoala commented:
I've scrap most of the old example in favour of a more simple and concise one. Also incorp everything into a working live example (tested locally).

yookoala committed:
Minor text change

* Better comment to describe

yookoala commented:
@chrisdavidmills: All issues have been addressed. Please check.

yookoala committed:
Fix linter issue with whitespace

chrisdavidmills commented on files/en-us/web/css/_colon_fullscreen/index.md:
````diff
@@ -23,44 +23,54 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
 }
 ```
 
+```js
````
A lead-in sentence would be nice here too, to break up the code blocks. How about

> The following JavaScript provides an event handler function that toggles fullscreen when the button is clicked.

chrisdavidmills reviewed (changes requested):
@yookoala much better, thanks! A few more comments to look at, but this is getting closer to the finish line.

yookoala commented on files/en-us/web/css/_colon_fullscreen/index.md:
````diff
@@ -23,44 +23,54 @@ The `:fullscreen` pseudo-class lets you configure your stylesheets to automatica
+});
+```
+
+### Result
````
It's strange because the :modal page also does the same thing (I'm simply copying the style here). Also things work locally on my computer. Could there be setting difference between the mdn content server from the development server?

yookoala committed:
Add leading statement to seperate the code blocks

* Add leading statements to CSS and Javascript code block for
  readability.

yookoala committed:
Downgrade "Result" heading to h4

yookoala commented:
Seems the `.requestFullscreen()` method call was blocked.

![image](https://github.com/mdn/content/assets/91274/91f53ca8-5f7d-49f4-b2da-bd47529acd3c)

yookoala commented:
Should I remove the live example embed?

chrisdavidmills commented:
> Seems the `.requestFullscreen()` method call was blocked.
> 
> ![image](https://user-images.githubusercontent.com/91274/279674374-91f53ca8-5f7d-49f4-b2da-bd47529acd3c.png)

@yookoala ah, so that's why it was failing. I should have checked the devtools ;-)

This is why the examples on other pages link to external examples. How about removing the live example, and just linking to a jsfiddle instead?

yookoala commented:
> > Seems the `.requestFullscreen()` method call was blocked.
> > 
> > ![image](https://user-images.githubusercontent.com/91274/279674374-91f53ca8-5f7d-49f4-b2da-bd47529acd3c.png)
> 
> @yookoala ah, so that's why it was failing. I should have checked the devtools ;-)
> 
> This is why the examples on other pages link to external examples. How about removing the live example, and just linking to a jsfiddle instead?

Sure. Can do that.

yookoala committed:
Use JSFiddle for demo instead of live example

* FeaturePolicy on mdn live example site forbid the use of
  Fullscreen API there. Need to place the demo on platform
  that allow the use of that.

yookoala commented:
@chrisdavidmills: Modified to use JSFiddle as demo platform.

chrisdavidmills committed:
Merge branch 'main' into fix/fullscreen-pseudo-class-docs

chrisdavidmills reviewed (approved):
@yookoala Great, this is looking really good! I'm just going to make a couple of small text adjustments, then I'll get it merged. Thanks so much for all your work on contributing to MDN!

chrisdavidmills committed:
Small text tweaks
